### ISSUE TITLE
[Bug]: Options Provider slow down open process of object

### ISSUE THREAD

**Issue by matthiashamacher** [pimcore/pimcore#10468]
### Expected behavior

Objects open within a reasonable timespan regardless of the configuration.

### Actual behavior

Opening an object takes more than 20 seconds when multiple option providers are configured and result in many possible options.

### For context
We previously had problems with option providers because they make the opening process much slower. Therefore we implemented caching into the option providers so they do not need to make all calls whenever the options are loaded. But since the update to Version 10.1.3, the caching didn't do anything, and opening an object with option providers takes more than 20 seconds.

### Steps to reproduce

1. Create multiple classes
2. Create a class with option providers who will give the other objects as options
3. Create an object of that last class and try to open it

**Version 10.1.3 or higher is needed**

**Comment by brusch** [comment]
Would be nice if you could debug this and provide a fix as a PR - since that's quite hard for us to reproduce and you have an actual case running 😊 Thanks!

**Comment by NiklasBr** [comment]
Agree that it is not optimal, it appears that the OptionProvider is called at least twice, sometimes more, for each field on opening an object. We partially worked around it with a static option:

```php
    public function getOptions($context, $fieldDefinition): array
    {
        // Using static storage because this method is called multiple times per dropdown per object
        static $options = [];

        if (!empty($options)) {
            return $options;
        }

        // Do stuff to populate options array

        return $options;
    }
```

**Comment by matthiashamacher** [comment]
Yeah, that's why we implemented caching. But since version 10.1.3 it doesn't seem to work anymore. 

My current guess is that it is because of this change: https://github.com/pimcore/pimcore/pull/10277 but I will have a closer look at it next week.

**Comment by brusch** [comment]
Yes, I think @NiklasBr proposed solution is best in that case, but I'd say this should be up to the developer to decide whether the options should be cached or not.

**Comment by matthiashamacher** [comment]
@brusch Why did you close the issue? 
I totally agree that it should be the decision of the developer what to do, but as we already use caching and it does not work anymore since 10.1.3 there is obviously a bug. Therefore this is still an open topic.

**Comment by brusch** [comment]
Can't see any issues with the cache, at least it's not reproducible for me. 
So from our side we not going to further work on this. I can reopen it, but we'd have to wait for a PR then.

**Comment by matthiashamacher** [comment]
I looked at this issue and found the change which causes the long loading time. It is this PR #10241 which is a fix for #10238 
After some debugging, I found out that it is caused because now there is the possibility that `$translationTerm` is empty when adding to the catalogue (https://github.com/pimcore/pimcore/blob/10.x/lib/Translation/Translator.php#L224). As soon as I change to the following everything works fine again, but the issue which should be fixed with that PR is also back again, so that is no valid solution:

**Before**
```php
if (empty($translation['type']) || $translation['type'] === 'simple') {
    $data[$translationKey] = $translationTerm;
} else {
    $dataIntl[$translationKey] = $translationTerm;
}
```

**After**
```php
if (empty($translation['type']) || $translation['type'] === 'simple') {
    if (empty($translationTerm)) {
        $data[$translationKey] = $translationKey;
    } else {
        $data[$translationKey] = $translationTerm;
    }
} else {
    $dataIntl[$translationKey] = $translationTerm;
}
```

From the PHP side the execution is relatively normal with the issue there must be something wrong with the js. So I had a look at the [Translate function](https://github.com/pimcore/pimcore/blob/10.x/bundles/AdminBundle/Resources/public/js/pimcore/functions.js#L13) and did some debugging:
At first, I thought it may be connected with this [part of the function](https://github.com/pimcore/pimcore/blob/10.x/bundles/AdminBundle/Resources/public/js/pimcore/functions.js#L63) that now all the missing translations are added to the "translation_admin_missing" array but after adding some `console.log` I could see that there are not many calls that go inside any of that if clauses (around 5 at max and only inside the first so they have a default value or are already translated.
Then I've added a `console.log` at the start of the function to see if it may be called less with the previous version, but that could not be confirmed (both times around 18000 calls per opening an object). But what I found out with that debugging is that before (without #10241  applied) the object is already shown after the first ~400 Calls of that function and the rest of the calls run in the background. After #10241  is applied the object is only shown after all 18000 Calls are finished, therefore the loading of the object is longer.

Has anybody an idea why the change from #10241 causes the javascript to first run all translation calls before opening the object? I currently have no idea where to look so maybe just give me a hint about what I can check and I can continue debugging.

**Comment by fashxp** [comment]
@dvesh3 maybe you have an idea?

**Comment by dvesh3** [comment]
@matthiashamacher Hi, could you please share your OptionsProvider implementation? It seems that you are using translator for option keys along with some caching, so we need to replicate the same scenario to debug the issue. thanks!

**Comment by matthiashamacher** [comment]
@dvesh3 Sure thing. You can find the cache service and an options provider in the following gist: https://gist.github.com/matthiashamacher/3e03541b7b04bbb12c7cf285256cc401
All other options providers are built the same just querying different tables and fields. We aren't using direct translator calls in the option provider, but the [calls come from the javascript](https://github.com/pimcore/pimcore/blob/10.x/bundles/AdminBundle/Resources/public/js/pimcore/object/tags/select.js#L238) where it tries to translate the keys of the option provider. These are ~3000 options for one and ~12000 options for another option provider. With some more option providers with max a few hundred options we get to ~18000 translation calls during the opening of an object.
Before the change from #10241 the object was opened after a few hundred calls, now the object only opens after all ~18000 translation calls are finished.

**Comment by matthiashamacher** [comment]
@dvesh3 Any updates on this?

**Comment by dvesh3** [comment]
@matthiashamacher Sorry, didn't get a chance to look at it, since we are preparing for Core and bundle releases scheduled for today. I will check this week or next.

**Comment by matthiashamacher** [comment]
@dvesh3 Any updates?

**Comment by dvesh3** [comment]
Sorry, not yet. I have assigned this to next bugfix milestone, so that i don't forget to check it 😉

**Comment by dvesh3** [comment]
@matthiashamacher I tried to reproduce the issue with 15000 calls to trans function and there 3 select types with option providers populating 5k options. I don't really see any difference in loading the object with or without changes from #10241. 

Could you please debug it further? have you tried returning true from `hasStaticOptions()`?

some other changes in 10.1.3 related to select datatype https://github.com/pimcore/pimcore/pull/10250

**Comment by matthiashamacher** [comment]
@dvesh3 Unfortunately setting the `hasStaticOptions()` to true does not work either. The loading time is the same.

Regarding the other changes, I figured out that the issue is because of #10241 because I was on version 10.1.2 and manually implemented the patches of each PR and tested again. Only with #10241 the loading time is increased, so I cannot reproduce the issue with #10250. Therefore I currently have no idea what I can debug further.

What I tested was to update to the latest version 10.2.5 and it seems that there the loading time is better. Unfortunately, I am getting another error regarding the classification store which then does not open the product at all.
The error only occurs in the 10.2 Versions and it seems that the name of the classification store key is not set correctly. Do you have an idea about this error so I can try on the newest version? Otherwise, I will need to have a deeper look at it in the next few days.

```
Pimcore\Model\DataObject\ClassDefinition\Data\Select::doEnrichDefinitionDefinition(): Argument #2 ($fieldname) must be of type string, null given, called in /var/www/pimcore/vendor/pimcore/pimcore/models/DataObject/ClassDefinition/Data/Select.php on line 466
```

**Comment by matthiashamacher** [comment]
@dvesh3 
I could locally fix my issue with the classification store but the loading time is still not better in the current version. 

But I had now the time to create some files with whom you can test and reproduce the whole issue. You can find them here: https://gist.github.com/matthiashamacher/cbb8e3cd8dfaa2bf89d80d142ec73315 
Those files represent the OptionsProvider Setup we use in the project where we have these problems. As mentioned before we have around 18000 Options in total in the project which are split throughout 8 OptionsProvider. The largest has around 13000 Options and the smallest has 10 Options.

If you have a look at the [Results](https://gist.github.com/matthiashamacher/cbb8e3cd8dfaa2bf89d80d142ec73315#results) for this issue, you can see that something changed in the time until an object is opened, if it has an option provider. All these numbers were timed with the Objects, OptionsProvider, Services, EventListener you find in the gist.

**Comment by dvesh3** [comment]
Thanks @matthiashamacher for detailing this out so well 👍 . I'll definitely have a look at it once we are done with 10.3 release.

**Comment by brusch** [comment]
To be honest, I don't think it's good to have 13000 options in one field ... I'm not even sure if this would work with a standard `<select>` in HTML?

**Comment by matthiashamacher** [comment]
I cannot tell you if it works in standard HTML, but it definitely works in Pimcore prior to version 10.1.3. As you can see from my linked gist we currently have a load time which is totally ok in version 10.1.2.

**Comment by CountZero1981** [comment]
the main slowdown occurs during DOM construction. I use this workaround: https://gist.github.com/CountZero1981/34426cbc02f1f99538e7236a318cac04. I inherited Pimcore select class, replace options before object creation and recreate them asynchronously after dashboard is rendered.
